crates/wasi-provider/src/wasi_runtime.rs

```diff
@@ -71,64 +88,246 @@ impl WasiRuntime {
 
         for (key, value) in self.dirs.iter() {
             let guest_dir = value.as_ref().unwrap_or(key);
-            // Try and preopen the directory and then try to clone it. This step adds the directory to the context
             ctx_builder_snapshot = ctx_builder_snapshot.preopened_dir(preopen_dir(key)?, guest_dir);
             ctx_builder_unstable = ctx_builder_unstable.preopened_dir(preopen_dir(key)?, guest_dir);
         }
         let wasi_ctx_snapshot = ctx_builder_snapshot.build()?;
         let wasi_ctx_unstable = ctx_builder_unstable.build()?;
 
-        let wasi_snapshot = Wasi::new(&store, wasi_ctx_snapshot);
-        let wasi_unstable = WasiUnstable::new(&store, wasi_ctx_unstable);
-        let module = wasmtime::Module::new(&store, &self.module_data)
-            .map_err(|e| format_err!("unable to load module data {}", e))?;
-        // Iterate through the module includes and resolve imports
-        let imports = module
-            .imports()
-            .iter()
-            .map(|i| {
-                // This is super funky logic, but it matches what is in 0.11.0
-                let export = match i.module() {
-                    "wasi_snapshot_preview1" => wasi_snapshot.get_export(i.name()),
-                    "wasi_unstable" => wasi_unstable.get_export(i.name()),
-                    other => bail!("import module `{}` was not found", other),
-                };
-                match export {
-                    Some(export) => Ok(export.clone().into()),
-                    None => bail!(
-                        "import `{}` was not found in module `{}`",
-                        i.name(),
-                        i.module()
-                    ),
+        // Clone the module data so it can be moved and we don't have to worry
+        // about the lifetime of the struct data
+        let module_data = self.module_data.clone();
```

Comment:
Does cloning here make more sense then wrapping the `module_data` in an `Arc`?

Reply:
I think I did it right. It required an `as_ref` while passing it to the function to satisfy the trait bound, but I could have done it wrong.